File: top_100_liked/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        found, ans = {}, []
        nums.sort()
        ### twoSums as subroutine
        for i, n in enumerate(nums):
            if(n in found):
                continue
            result = self.twoSum(nums[i+1:], 0 - n)
            found[n] = 1
            for l in result:
                ans.append((n,*l))
        return ans

    def twoSum(self, nums, target):
        mapp, found, result = {}, {}, []
        for n in nums:
            if(n in mapp and n not in found):
                result.append([n, target-n])
                found[n] = 1
            mapp[target-n] = 1 # dummy number
        return result

File: top_100_liked/cpp/3sum.cpp
class Solution {
public:
    // size of nums, range of nums[i]
    // use two sums as subroutine
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int> > result;
        int i = 0, n = nums.size();
        while(i < n){
            // early stopping
            if(nums[i] > 0)
                break;
            twoSum(nums, nums[i], i+1, n-1, -nums[i], result);
            int prev = nums[i];
            while(i < n && nums[i] == prev)
                i++;
        }
        
        return result;
    }
    
    void twoSum(vector<int>& num, int k, int start, int end, int target, vector<vector<int> >& result){
        int l = start, r = end;
        
        while(l < r){
            int sum = num[l] + num[r];
            if(sum < target)
                l++;
            else if(sum > target)
                r--;
            else{
                result.push_back({k, num[l], num[r]});
                int prev = num[l];
                while(l < r && num[l] == prev)
                    l ++;
            }
        }
    }
};
